### PR TITLE
Warn if user is paying, but are already paid up > 2 months (too soon)

### PR DESCRIPTION
**Pull request by weedySeaDragon**
#### PT URL: https://www.pivotaltracker.com/story/show/166273184

In order to determine if it is "too soon" for a user to pay and then show them a warning and info in the UI, we need to know (1) how many days is "too soon", and then (2) use that to figure out if a user 'should' make a payment now and is allowed to make the payment.

We need to confirm these assumptions that I implemented (@thesuss, @patmbolger) :
- [x] a user is _allowed to pay a membership fee_ IF the user is _not_ an admin AND either (a) is a member, OR (b) is a user and has at least one application with status == accepted
- [x] a user is _allowed to pay a branding fee_ IF (a) the user is an admin, OR (b) the user is a member _and_ is in the company
- [x] an admin _can_ pay a branding fee (_as an admin_) (see the feature added to features/payments_and_branding_status).  An admin can still _become a user_ and pay the branding fee _as the user._
- [x] an admin _CAN_ pay a membership fee (_as an admin_) (see the feature added to features/payments_and_membership).  An admin can still _become a user_ and pay the membership fee _as the user._  **This will be a separate PR.**  (see PT: https://www.pivotaltracker.com/story/show/169765762)
- [x] a payment _should_ be made now IF (a) the term has expired, OR (b) it is not "too soon"
    - too soon = if today is _on_ or after the 'too soon cutoff date' [cutoff_date = expiration date -  'too soon' number of days.  
    - Ex: expiration date = 31 December, "too soon number of days" = 60 days, then the "too soon cutoff date" = 31 December - 60 days = 1 November.   
      It is "too soon" if the user tries to pay on 1 November.  It is _not_ too soon if the user tries to pay on 2 November.
    - see the block comment in `model/concerns/payment_utility.rb` for `should_pay_now?` for detailed examples

The number of days that is considered "too soon" is set in the Application Configuration (AppConfig) info.  (default = 60 days).

---
### When an expiration date is known (= when payments have been made), a "... due by" hint is shown: 

membership-term-payment due hint-sv-due-by75
---

membership-term-payment due hint-en-due-by75
---
---

### "past due" hint shown when an expiration date (= due date) is known:

membership-term-payment due hint-sv-past-due75
---
membership-term-payment due hint-en-past-due75

---
---
### "too early" hint:

membership-term-payment due hint-sv-too-early75

---

membership-term-payment due hint-en-too-early75

---
---
### The same messages show for the H-Branding fee for a Company.
  Here's a screenshot showing the "due by..." hint::

h-markt-term-payment due hint-sv-due-by75

---
---

### "Payment is due now" for an Approved Applicant
When the first payment is due, "Payment is due now" is shown as the hint under the button.

not-member-payment-due-now-sv75x

---
---
### Additions to the locales/sv.yml file:
```
  close: Stänga
  lastupdated: &last_updated Senast uppdaterad

  membership: &membership medlemskap


  activerecord:
    attributes:
      app_configuration:
          payment_too_soon_days: &payment_too_soon_days Number of days that a user will be warned that a payment is 'too soon'


  payors:
    status: Status
    no_payment_due_now: &no_payment_due_now Ingen betalning behöver göras nu.
    due: Betalning sker nu.
    past_due: Betalning är förfallen!
    due_by: Betalning sker före %{due_date}.
    too_early: *no_payment_due_now
    term_paid_through: &term_paid_through Betald till och med
    you_are_paid_until: Du betalas tills
    you_dont_owe_until_then: Du är inte skyldig en betalning förrän då.
    paying_now_extends_until: "Att göra en %{fee_name_payment} nu kommer att förlänga din %{term_name} till %{extended_end_date}."
    do_you_want_to_pay_now: Vill du verkligen göra en betalning nu?
    yes_pay_answer: 'Ja, gör en betalning'
    no_pay_answer: 'Nej. Nevermind. Betal inte'
    admin_cant_edit: "Medlemsstatus kan inte uppdateras. (Kanske har inga betalningar gjorts.)"


  payment:
    processor_status: Processor Status
      status_from_payment_processor:
        skapad: skapad
        avvaktan: avvaktan
        betald: betald
        utgånget: utgånget
        väntar_på_betalning: "Väntar på betalning"
      term_start_date: startdatum för terminen
      term_expiry_date: slutdatum
      no_payment_due_now: *no_payment_due_now


  companies:
    show:
      hbrand_status_paid: Kan använda H-märkesbilden. (Licensavgiften betalas)
      hbrand_status_not_paid: Kan inte använda H-märkesbilden. (Licensavgiften betalas inte)
      hbrand_term_payment: betalning för H-märkespremie licensperiod
      hbrand_fee: avgift för H-märkespremie
      hbrand_term: licensperiod för H-märkespremie
      payment_tooltip: Betald avgift förlänger H-märkningen i ett år.


  users:
    show:
      membership_status: Medlemsstatus
      not_a_member: Äar inte medlem
      is_a_member: Är en medlem med god status
      term_paid_through: *term_paid_through
      no_payment_due_now: *no_payment_due_now
      membership_fee_payment: medlemsbetalning
      membership_fee: medlemskaps avgift
      membership_term: medlemstid
      payment_tooltip: När du betalar din medlemsavgift förlängs ditt medlemskap i ett år.

      edit_member_status: &edit_member_status Redigera medlemsstatus
      pay_membership: &pay_membership Betala medlemsavgift



```

---

I tried make reviewing a bit easier by adding words like "(minor)" to the commit names

---

### Model and main changes
1.  add the number of days to be considered "too soon to pay" in **AppConfiguration** (includes db change, additions to the AppConfiguration `show` and `edit` views)

2.  added methods to **User** so we can determine:
    1. Is a user allowed to make a membership payment?
    2.  Is a user allowed to make a branding license fee payment?
    3. is the user "in" a company? 
        - does the user have an approved application that contains/has a particular company?

3. concerns/PaymentUtility
   - added ‘Has the term expired?
   - added "should a payment be made now?", "is it too soon to make a payment?" (which is just the opposite (_not_) of "should a payment be made now?"
   - added payment_type = self.class::THIS_PAYMENT_TYPE to methods so that by default the payment type is set by the containing class (e.g. User or Company).

4.  an admin _can_ pay a branding fee (as an admin) (added feature to features/payments_and_branding_status)

5.  an admin _cannot_ pay a membership fee (added feature to features/payments_and_membership)

### UX
- refactored common elements into "payor" an entity payments are made for. e.g. User, Company
- use "term" to encompass all info about the term (start, expire dates, status, payment due information) (‘payment’ is more specific information; it applies to a term)
companies/branding_payment_status.html.haml -> branding_term_status.html.haml
users.membership_payment_status.html.haml -> membership_term_status.html.haml

- Use CSS classes to display ‘yes’ ‘no’ ‘maybe’ ‘admin-only’  This can be leveraged/improved on with future PRs. 

## Screenshots (Optional):
see the screenshots for #710 - the UX for this

I did change how the admin buttons for "Redigera medlemsstatus"/"Edit membership status" and "Redigera Status"/"Edit Branding Status" are displayed to fix the blue background/div area overlap problem.  I just moved those buttons down into their own CSS row.

---
## Ready for review:
@patmbolger 
@thesuss 
@AgileVentures/shf-project-team

### CODE REVIEW

**Comment by thesuss** [comment]
@weedySeaDragon 
_an admin cannot pay a membership fee (as an admin) (see the feature added to features/payments_and_membership)_ 
**Is false.** 
Admin **can** pay for another user, I did that for our mutual person the other day. 

Aha, now I realize this hindrance is a new thing in this PR :) 
How to do next year then? Change to that user and go that way - of course 👍

**Comment by patmbolger** [comment]
Just started looking at this PR.  First comments:

1. On @thesuss comment above - yes, the admin can _become_ a user and pay the membership fee for that user.  But how will the payment be made?  If using the member's credit card (cc), then the member will have to provide the credentials to the admin.  If using the admin's own cc, then HIPS will remember that credit card and associate it with the member's email (assuming the member's email is used for the payment).  Not sure of this will work smoothly ....

2. "a user is allowed to pay a membership fee IF the user is not an admin AND either (a) is a member, OR (b) is a user and has at least one application with status == accepted" - please note that a member has only _one_ application (this was refactored from _has_many_ a while ago).

**Comment by thesuss** [comment]
@patmbolger  not CC, invoice, no hassle :)

File: app/helpers/payments_helper.rb [hunk 1]
```diff
@@ -4,32 +4,63 @@ module PaymentsHelper
...
+
+
+  def payment_should_be_made_class(entity)
+    # TODO use the classes defined in ApplicationHelper
+    if entity.term_expired?
+      'No'
+    elsif entity.too_early_to_pay?
+      'Yes'
+    else
+      'Maybe'
```
Comment (patmbolger) [inline app/helpers/payments_helper.rb]: I know this is just the CSS class that we use to indicate whether the user should (or _maybe_ should) pay a fee or not.  However, the method name is confusing when coupled with the return values.

Perhaps a name like `payment_status_ok_indicator_css_class` would be better(?).  And, yes, maybe too wordy.

File: app/helpers/payments_helper.rb [hunk 2]
```diff
@@ -4,32 +4,63 @@ module PaymentsHelper
...
 
-    unless expire_date
-      return field_or_none("#{t('activerecord.attributes.payment.expire_date')}",
-                           "#{t('none_t')}", label_class: 'standard-label')
+
+  # @param entity [User or Company] - the entity with a possible membership expiration date
+  # @return [nil | date] - return nil if there is no expiration date (e.g. not a member), else
+  # the Date that the current membership term expires
+  # TODO - this is a smell (checking the class): polymorphism should be used instead. Default implementation should be in a parent class for User and Company
+  def entity_expire_date(entity = nil)
```
Comment (patmbolger) [inline app/helpers/payments_helper.rb:35]: Not sure I agree with this statement.  User and Company are quite distinct, and making them subclasses of a parent (Payor?) class would be a mistake, IMO.  The two classes would be subclassed based on a common attribute, not based on being a "type of" or specific examples of the parent class.

Comment (weedySeaDragon) [reply app/helpers/payments_helper.rb:35]: Agree that User and Company have a lot of differences.

But they do definitely share the common behavior of making and handling _payments_.
This behavior is what I think of as a `Payor`   Refactoring out this behavior, and then refactoring out the different `Payments` is what we should discuss and work through.

File: app/helpers/payments_helper.rb (hunk 1 above)
Comment (weedySeaDragon) [reply app/helpers/payments_helper.rb]: excellent catch.  Made me realize it would be helpful to always make sure it's clear when we're dealing with _CSS classes_ vs. "normal" Ruby classes. 
I'll hit up the thesaurus to see if I can come up with something more succinct.  (Which is not my strong suit ;-) )
This also prompted me to enter a small chore story in PT to do some refactoring so we can simplify and standardize the 'yes/no/maybe' CSS classes and usage.

**Comment by patmbolger** [comment]
Here is a screenshot showing the user's account when the membership payment is good for at least another 2 months:

Screen Shot 2019-12-04 at 3 50 37 PM

The message under the payment button is very helpful.

However, there is no such message when we're within 2 months of expiration:

Screen Shot 2019-12-04 at 3 51 55 PM

nor when the payment has expired:

Screen Shot 2019-12-04 at 3 53 52 PM

I think a message under the button is even _more_ important in these last 2 scenarios than the first.

I recommend adding such a message(s), and style the text background consistent with the "maybe" and "no" styling of the user's paid-through status.

**Review by patmbolger — CHANGES_REQUESTED** [review]
Looks good but please see comment.

**Review by patmbolger — APPROVED** [review]
Looks great!

**Comment by thesuss** [comment]
close: Stäng
(if it is like close a prompt) 

    due: Betalning behöver göras.
    due_by: Betalning behöver göras före %{due_date}.
    you_are_paid_until: Du har betalt till och med
    you_dont_owe_until_then: Du är inte skyldig en betalning förrän:
    no_pay_answer: 'Nej.'
      term_start_date: startdatum

      hbrand_status_paid: Får använda H-märkesbilden. (H-märkespremie betalad)
      hbrand_status_not_paid: Får inte använda H-märkesbilden. (H-märkespremie ej betad)
      hbrand_term_payment: H-märkespremiebetalning
      hbrand_fee: H-märkespremie
      hbrand_term: H-märkespremieperod
      payment_tooltip: Betald avgift förlänger H-märkningen i ett år.

      not_a_member: Är inte medlem
      is_a_member: Är medlem 
      membership_fee_payment: betalning av medlemskap
      membership_fee: medlemskapsavgift
      membership_term: medlemperiod
